Gordon360/Services/MembershipService.cs

```diff
@@ -88,167 +92,108 @@ public MEMBERSHIP GetSpecificMembership(int membershipID)
         /// </summary>
         /// <param name="activityCode">The activity code.</param>
         /// <param name="sessionCode">Optional code of session to get memberships for</param>
+        /// <param name="groupAdmin">Optional filter for group admins</param>
+        /// <param name="participationTypes">Optional filter for involvement participation type (MEMBR, ADV, LEAD, GUEST)</param>
         /// <returns>MembershipViewModel IEnumerable. If no records were found, an empty IEnumerable is returned.</returns>
-        public async Task<IEnumerable<MembershipViewModel>> GetMembershipsForActivityAsync(string activityCode, string? sessionCode = null)
-        {
-            IEnumerable<MEMBERSHIPS_PER_ACT_CDEResult> memberships = await _context.Procedures.MEMBERSHIPS_PER_ACT_CDEAsync(activityCode);
-
-            if (sessionCode != null)
-            {
-                memberships = memberships.Where(m => m.SessionCode.Trim() == sessionCode);
-            }
-
-            return memberships.OrderByDescending(x => x.StartDate).Select(m => new MembershipViewModel
-            {
-                MembershipID = m.MembershipID,
-                ActivityCode = m.ActivityCode.Trim(),
-                ActivityDescription = m.ActivityDescription.Trim(),
-                ActivityImagePath = m.ActivityImagePath,
-                SessionCode = m.SessionCode.Trim(),
-                SessionDescription = m.SessionDescription.Trim(),
-                IDNumber = m.IDNumber,
-                AD_Username = m.AD_Username,
-                FirstName = m.FirstName.Trim(),
-                LastName = m.LastName.Trim(),
-                Mail_Location = m.Mail_Location,
-                Participation = m.Participation.Trim(),
-                ParticipationDescription = m.ParticipationDescription.Trim(),
-                StartDate = m.StartDate,
-                EndDate = m.EndDate,
-                Description = m.Description,
-                GroupAdmin = m.GroupAdmin,
-                AccountPrivate = m.AccountPrivate
-            });
+        public IEnumerable<MembershipView> GetMembershipsForActivity(
+            string activityCode,
+            string? sessionCode = null,
+            bool? groupAdmin = null,
+            string[]? participationTypes = null
+
+        ) {
+            //IEnumerable<MEMBERSHIPS_PER_ACT_CDEResult> memberships = await _context.Procedures.MEMBERSHIPS_PER_ACT_CDEAsync(activityCode);
+
+            if (sessionCode == null)
+                sessionCode = this.GetCurrentSessionCode();
+
+            return _context.MembershipView
+                .Where(m => 
+                    m.SessionCode.Trim() == sessionCode
+                    && m.ActivityCode == activityCode
+                    && (groupAdmin == null || m.GroupAdmin == groupAdmin)
+                    && (participationTypes == null || participationTypes.Contains(m.Participation))
+                 )
+                .OrderByDescending(m => m.StartDate);
```

Comment:
This `Where` filter is complex enough that EF Core won't be able to translate it to SQL I suspect, which will force client evaluation. I think we could split it up and build up the IEnumerable so that it gets evaluated in SQL instead. Something like:
```suggestion
            IEnumerable<MembershipView> memberships =  _context.MembershipView
                .Where(m => 
                    m.SessionCode.Trim() == sessionCode
                    && m.ActivityCode == activityCode);
           if (groupAdmin is not null) {
               memberships = memberships.Where(m => m.GroupAdmin == groupAdmin);
           }
           
           if (participationTypes?.Length > 0) {
               memberships = memberships.Where(m => participationTypes.Contains(m.Participation));
           }
           return memberships.OrderByDescending(m => m.StartDate);
```

That may still be complex enough to force client evaluation, just because of the `Trim()` and `Array.Contains` calls (I'm not sure if EF Core is smart enough to translate those to SQL), but it's worth a shot.

Reply:
That makes sense, thanks. I think the first part might be fine. It's likely EF Core is able to translate *.Trim() to TRIM(*) in SQL.